packages/iov-bns/src/bnsconnection.ts

```diff
@@ -190,55 +193,68 @@ export class BnsConnection implements BcpAtomicSwapConnection {
     const transactionId = Encoding.toHex(postResponse.hash).toUpperCase() as TransactionId;
 
     // can be undefined as we cannot guarantee it assigned before the caller unsubscribes from the stream
+    let transactionSubscription: Subscription | undefined;
     let blockHeadersSubscription: Subscription | undefined;
 
-    const firstEvent: BcpBlockInfo = { state: BcpTransactionState.Pending };
-    let lastEventSent: BcpBlockInfo = firstEvent;
-    const blockInfoProducer = new DefaultValueProducer<BcpBlockInfo>(firstEvent, {
-      onStarted: async () => {
-        try {
-          // we utilize liveTx to implement a _search or watch_ mechanism since we do not know
-          // if the transaction is already committed when the producer is started
-          const searchResult = await toListPromise(this.liveTx({ id: transactionId }), 1);
-          const transactionHeight = searchResult[0].height;
-          const transactionResult = searchResult[0].result;
-
-          // Don't do any heavy work (like subscribing to block headers) before we got the
-          // search result. For some transactions this will never resolve.
-
-          {
-            const inBlockEvent: BcpBlockInfoInBlock = {
-              state: BcpTransactionState.InBlock,
-              height: transactionHeight,
-              confirmations: 1,
-              result: transactionResult,
-            };
-            blockInfoProducer.update(inBlockEvent);
-            lastEventSent = inBlockEvent;
-          }
-
-          blockHeadersSubscription = this.watchBlockHeaders().subscribe({
-            next: async blockHeader => {
-              const event: BcpBlockInfo = {
-                state: BcpTransactionState.InBlock,
+    const firstEvent: BlockInfo = { state: TransactionState.Pending };
+    let lastEventSent: BlockInfo = firstEvent;
+    const blockInfoProducer = new DefaultValueProducer<BlockInfo>(firstEvent, {
+      onStarted: () => {
+        transactionSubscription = this.liveTx({ id: transactionId }).subscribe({
+          next: searchResult => {
+            if (isFailedTransaction(searchResult)) {
```

Comment:
This is the core of this PR. Due to an improved liveTx that can handle failed transactions, postTx now supports propagating delivertx errors.